src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+                                    foreachInfo.RequireExplicitCast ?
+                                        SyntaxFactory.CastExpression(
+                                            SyntaxFactory.ParseTypeName(foreachInfo.ExplicitCastInterface).WithAdditionalAnnotations(Simplifier.Annotation),
+                                            SyntaxFactory.ParenthesizedExpression(foreachCollectionExpression)) : 
+                                        foreachCollectionExpression)))));
+
+                collectionStatement =
+                    collectionStatement.ReplaceToken(collectionStatement.GetFirstToken(), collectionStatement.GetFirstToken().WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+
+                editor.InsertBefore(foreachStatement, collectionStatement);
+            }
+
+            // create new index varialbe name
+            var indexString = CreateUniqueMethodName(model, foreachStatement.Statement, "i");
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(foreachInfo, parseOption, collectionVariableName, indexString);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(SyntaxFactory.IdentifierName("var"),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                        SyntaxFactory.Identifier(indexString).WithAdditionalAnnotations(RenameAnnotation.Create()),
+                        default,
+                        SyntaxFactory.EqualsValueClause(
+                            SyntaxFactory.LiteralExpression(SyntaxKind.NumericLiteralExpression, SyntaxFactory.Literal(0)))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                SyntaxFactory.BinaryExpression(SyntaxKind.LessThanExpression, SyntaxFactory.IdentifierName(indexString), SyntaxFactory.ParseExpression($"{collectionVariableName}.{foreachInfo.CountName}")),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(SyntaxFactory.PostfixUnaryExpression(SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexString))),
+                bodyStatement);
+
+            // let leading and trailing trivia set
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                forStatement = forStatement.ReplaceToken(forStatement.ForKeyword, forStatement.ForKeyword.WithLeadingTrivia(foreachStatement.ForEachKeyword.LeadingTrivia));
+            }
+
+            forStatement = forStatement.ReplaceToken(forStatement.CloseParenToken, forStatement.CloseParenToken.WithTrailingTrivia(foreachStatement.CloseParenToken.TrailingTrivia));
+
+            editor.ReplaceNode(foreachStatement, forStatement);
+        }
+
+        private ExpressionSyntax GetForeachCollection(SyntaxNode node)
```

Comment:
This seems odd that node is typed as SyntaxNode and not ForEachStatementSyntax.  And once you have that, this whole method seems odd to have.

Reply:
since ForeachInfo is type in Core and will be used for both C# and VB

Reply:
That does not mean you cannot be strongly typed.  Please look at virtually any other refactoring/analyzer.  When we work across the languages and we operate with syntax nodes, we do so in a strongly typed fashion.  please follow the patterns and practices of the majority of the IDE code actions.